src/biokbase/narrative/jobs/appmanager.py

```diff
@@ -466,41 +415,281 @@ def run_app(
-        return -1
+        # make sure we have all required keys
+        required_keys = ["app_id", "tag", "params"]
+        for key in required_keys:
+            if key not in app_info:
+                raise ValueError(
+                    f"app info must contain keys {', '.join(required_keys)}"
+                )
+        # make sure app is of the form "module/app"
+        if (
+            not isinstance(app_info["app_id"], str)
+            or re.match(r"\S+\/\S+", app_info["app_id"]) is None
+        ):
+            raise ValueError("an app_id must be of the format module_name/app_name")
+        # params must be a list with at least one item (even an empty dict)
+        if not isinstance(app_info["params"], list) or len(app_info["params"]) == 0:
+            raise ValueError(malformed_params_error)
+        # each item must be a dict
+        for params in app_info["params"]:
+            if not isinstance(params, dict):
+                raise ValueError(malformed_params_error)
+        # make sure tag is an allowed item
+        allowed_tags = ["release", "beta", "dev"]
+        if app_info["tag"] not in allowed_tags:
+            raise ValueError(
+                f"tag must be one of {', '.join(allowed_tags)}, not {app_info['tag']}"
+            )
+        # make sure version is a string, if present
+        if "version" in app_info and not isinstance(app_info["version"], str):
+            raise ValueError(
+                f"an app version must be a string, not {app_info['version']}"
+            )
 
-        # typedef structure {
-        #     string method;
-        #     string app_id;
-        #     list<UnspecifiedObject> params;
-        #     string service_ver;
-        #     list<wsref> source_ws_objects;
-        #     Meta meta;
-        #     int wsid;
-        #     string parent_job_id;
-        # } RunJobParams;
+    def _build_run_job_params(
+        self,
+        spec: dict,
+        tag: str,
+        param_set: dict,
+        version: str = None,
+        cell_id: str = None,
+        run_id: str = None,
+        ws_id: int = None,
+    ) -> dict:
+        """
+        Builds the set of inputs for EE2.run_job and EE2.run_job_batch (RunJobParams) given a spec
+        and set of inputs/parameters.
+
+        This returns a dict with the following keys:
+            method: the function to run
+            service_ver: the version of the app to run
+            params: the set of inputs, mapped to what the method expects to see
+            app_id: the original id of the app to run
+            wsid: the workspace id associated with the new job
+            source_ws_objects: the UPAs for any workspace objects involved with running the app, if any
+            meta: key-value pairs, usually containing:
+                cell_id (if not None),
+                run_id (if not None),
+                tag
 
-        # typedef structure {
-        #     int wsid;
-        # } BatchParams;
+        Parameters:
```

Comment:
could you put the parameters first and the return structure afterwards?

Reply:
Sure, done.